core/src/callback.rs

```diff
@@ -33,16 +38,29 @@ mod callbacks_ptr {
     }
 }
 
-// `Self: Sync` because `F` is called from some zig thread.
+/// Reply returned by [`Callbacks`].
+pub struct Reply<'a> {
+    /// Returned raw payload of this [`Reply`] as bytes.
+    pub payload: &'a [u8],
+
+    /// Cluster timestamp when the reply was generated.
+    pub timestamp: SystemTime,
+}
+
+// `Self: Sync` because `F` is called from some Zig thread.
 pub trait Callbacks: Sync {
     type UserDataPtr: UserDataPtr;
 
-    fn on_completion(&self, packet: Packet<'_, Self::UserDataPtr>, payload: &[u8]);
+    /// Calls back once a [`Packet`] is submitted and processed.
+    ///
+    /// [`None`] `reply` means that submitting the [`Packet`] failed (check the [`Packet::status`]
+    /// for the reason).
```

Comment:
I would have prefered `Result`, maybe? Anyway this is a small change and given this is `core` it's not very relevant.

Reply:
@zetanumbers `Result` will duplicate the error, since it, if happens, is written into the referred `Packet::status`. Before `Option` there was just a slice, and instead of `None` the empty slice was returned, which is OK. But now we also return the `timestamp` and returning `0` for it seems quite wrong. I'll think how to refactor this, but it's beyond this PR, for sure.